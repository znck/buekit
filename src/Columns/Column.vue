<script>
import { cssModule, mergeData, validate, isResponsiveType, toArray, toMap, isString, isNumber, isUndefined } from '../utils'
import { tag } from '../mixins'

const sizes = toMap(['three-quarters', 'two-thirds', 'half', 'one-third', 'one-quarter'])

const _isNumber = /^[\d]+$/
function parseIntegerIfPossible (any) {
  if (isString(any) && _isNumber.test(any)) return parseInt(any)

  return any
}

function prepareSize (value) {
  const index = value.lastIndexOf('-')
  const device = value.substring(index + 1)

  if (isResponsiveType(device)) return { device, size: parseIntegerIfPossible(value.substring(0, index)) }

  return { size: parseIntegerIfPossible(value) }
}

function isValidGrid (value) {
  value = parseIntegerIfPossible(value)

  return isNumber(value) && value > 1 && value <= 12
}

function isValidSize (value) {
  value = prepareSize(value)
  
  return (
    (value.size in sizes || isValidGrid(value)) // validate size.
    && (isUndefined(value.device) || isResponsiveType(value.device)) // validate device.
  )
}

function createSizeType () {
  return {
    default: () => [],
    type: [Array, String, Number],
    validate: value => validate(value, isValidSize)
  }
}

export default {
  name: 'Column',
  functional: true,
  mixins: [tag],
  props: {
    narrow: {
      default: false,
      type: Boolean
    },
    /**
     * Take only required space.
     */
    narrowOnDevice: {
      default: () => [],
      type: [Array, String],
      validate: value => validate(value, isResponsiveType)
    },
    /**
     * Offset on left.
     */
    offset: createSizeType(),
    /**
     * Width of the column.
     */
    size: createSizeType(),
  },

  render (h, ctx) {
    const s = cssModule(ctx.$style)
    const classes = [ s('column') ]
    const push = (values, prefix) => toArray(values).forEach(
      value => { classes.push(s(prefix + '-' + value)) }
    )

    const { narrow, narrowOnDevice, size, offset, tag } = ctx.props

    narrow && classes.push(s(`is-narrow`))
    push(narrowOnDevice, 'is-narrow')
    push(size, 'is')
    push(offset, 'is-offset')
    
    return h(tag, mergeData({ class: classes }, ctx.data), ctx.children)
  }
}
</script>

<style lang="scss" src="./columns.scss" module></style>
