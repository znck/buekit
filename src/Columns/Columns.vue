<script>
import { validate, isNumber, isBoolean, isResponsiveType, toArray, toMap, mergeData, cssModule } from '../utils'
import { tag } from '../mixins'

const horizontally = toMap(['horizontally', 'both'])
const vertically = toMap(['vertically', 'both'])
const _isGap = /^[0-8]$/

export default {
  name: 'Columns',
  functional: true,
  mixins: [tag],
  props: {
    centered: {
      default: false,
      type: [Boolean, String],
      validate: value => isBoolean(value) || (value in horizontally || value in vertically)
    },

    device: {
      default: '',
      type: String,
      validate: value => value === '' || isResponsiveType(value)
    },

    gap: {
      default: true,
      type: [Boolean, Number],
      validate (value) {
        return _isGap.match(value) || isBoolean(value)
      }
    },

    multiline: {
      default: false,
      type: Boolean
    }
  },
  
  render (h, ctx) {
    const s = cssModule(ctx.$style)
    const styles = [s('columns')]

    const { gap, multiline, device, centered, tag } = ctx.props

    // multiline
    if (multiline) styles.push(s('is-multiline'))

    // responsiveness
    if (device) styles.push(s('is-' + device))
    
    // gap
    if (gap === false) {
      styles.push(s('is-gapless'))
    } else if (gap !== true) {
      styles.push(s('is-variable'))
      styles.push(s('is-' + gap))
    }

    // centered
    if (centered in horizontally || centered === true || centered === '') styles.push(s('is-centered'))
    if (centered in vertically) styles.push(s('is-vcentered'))
    
    return h(tag, mergeData({ class: styles }, ctx.data), ctx.children)
  }
}
</script>

<style lang="scss" src="./columns.scss" module></style>
