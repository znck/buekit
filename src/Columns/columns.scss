@import '~bulma/sass/utilities/_all.sass';
@import '../_base.scss';
@import '~bulma/sass/grid/columns.sass';
