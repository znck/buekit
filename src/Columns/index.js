import Columns from './Columns.vue'
import Column from './Column.vue'

export {
  Column,
  Columns
}

export const Row = Columns // Export as row for better api.
