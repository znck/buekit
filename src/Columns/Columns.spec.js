import { shallow } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'

import Columns from './Columns.vue'

function snapshots (options) {
  options.forEach(
    option => it ('matches snapshot: ' + JSON.stringify(option), () => {
      const wrapper = shallow(Columns, { props: option, slots: { default: '<p>content</p>' } })
      const renderer = createRenderer()
  
      renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) throw new Error(err)
        expect(str).toMatchSnapshot()
      })
    })
  )
}

describe('Columns.vue', () => {
  snapshots ([
    {},
    { centered: true },
    { centered: 'vertically' },
    { gap: false },
    { gap: 8 },
    { multiline: true }
  ])
})
