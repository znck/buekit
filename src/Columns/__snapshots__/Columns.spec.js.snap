// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns.vue matches snapshot: {"centered":"vertically"} 1`] = `<!---->`;

exports[`Columns.vue matches snapshot: {"centered":true} 1`] = `<!---->`;

exports[`Columns.vue matches snapshot: {"gap":8} 1`] = `<!---->`;

exports[`Columns.vue matches snapshot: {"gap":false} 1`] = `<!---->`;

exports[`Columns.vue matches snapshot: {"multiline":true} 1`] = `<!---->`;

exports[`Columns.vue matches snapshot: {} 1`] = `<!---->`;
